packages/model-viewer/src/three-components/gltf-instance/VariantMaterialLoaderPlugin.ts

```diff
@@ -143,7 +143,8 @@ export default class GLTFMaterialsVariantsExtension implements
           if (!extensionsDef || !extensionsDef[this.name]) {
             continue;
           }
-          meshes[i].userData.variantMaterials =
+          const meshObjectIndex = i < meshes.length  ? i : 0;

```

Comment:
This feels like a hack. I think you have discovered a real issue, but I'd like to fix it properly. Do you understand yet what circumstances lead to getting an out-of-bounds value?

Reply:
Agreed. Sorry about that. 

Here is quick log for the execution inside the traverse callback for my test model.

![Screen Shot 2022-02-10 at 10 44 48 PM](https://user-images.githubusercontent.com/8660478/153534534-e9f76dbe-ed4d-42f4-8ac4-980205549c92.png)

So when we traverse the scene recursively, we sometimes execute the code above for `Group` objects with `Mesh` objects(each representing a primitive of the original Mesh) as children and, then, we will execute for each of these `Mesh` object individually.  So that means that we actually assign `userData.variantMaterials` to each `Mesh` twice because of the nature of the traverse method. So using the index of the primitives array in the loop won't always work for accessing the corresponding mesh, since the `meshes` and `primitivesDef` arrays are only really corresponding when the object is a `Group` and we assign meshes to `object.children`. 

I think the loop could probably be completely removed here. We have all the necessary information in the `association` property of the `Mesh` objects to assign userData.variantMaterials without it. Also, in my case it seems that the last primitive's extension will be assigned to each mesh, which I didn't noticed before since the model and its variants is visually being rendered properly without any `userData.variantMaterials` whatsoever.

Does that make sense?

Reply:
That sounds reasonable to me; can you go ahead and make those changes in this PR (and remove the comment if you're not sure it applies anymore)? I'd like to test it and see it in context.

Reply:
We already have some variants regressions tests and if you add your new one as well, I'll feel fairly confident if everything passes.

Reply:
You're talking about the regressions tests in `render-fidelity-tools` right?

Reply:
No, I'm talking about these tests (and others) https://github.com/google/model-viewer/blob/master/packages/model-viewer/src/test/features/scene-graph-spec.ts#L88